Creating the keyset.yaml file if it does not exist

chrislovecnm opened the pull request:
If the Keypair is not found kops creates a new keyset file.  We are
setting the Format to v1, which denotes that we do not have a keyset
file.

This commit enables upgrading from kops 1.8 -> 1.9 while upgrading an
existing cluster.  Clusters built with kops 1.8 do not have the keyset
file, and these code changes allow the creation of the keyset file.

Fixes: https://github.com/kubernetes/kops/issues/4546

~~The only thing I do not like about this fix is that we are generating a new private key.  Not a horrible thing, but not sure how stuff like etcd will behave.~~

Edit: with changes I am not generating a new private key.

chrislovecnm commented:
/assign @blakebarnett @mikesplain

blakebarnett commented:
Apparently this results in a new private key being generated for the CA? This would require us to rotate every cert we've created using the CA (a large number), if there's any way to make it re-use the existing private key that would be very preferable.

cc: @justinsb

chrislovecnm commented:
@blakebarnett I am not certain that a new private cert is being generated or we are creating two private certs with a keyset.yaml.

@justinsb any ideas?

chrislovecnm commented on upup/pkg/fi/fitasks/keypair.go:
```diff
@@ -180,15 +185,22 @@ func (_ *Keypair) Render(c *fi.Context, a, e, changes *Keypair) error {
 	}
 
+	// TODO: I am wondering if we reuse the certificate information when a.KeysetAPIVersion == "" rather than
+	// TODO: issuing the new cert
```
I am seeing two private pems in the keypair.yaml file.  Any idea how we can re-use the pem?

chrislovecnm replied:
I think that we are eventually calling https://github.com/kubernetes/kops/blob/master/pkg/pki/csr.go#L94

And that is creating the new cert.

chrislovecnm replied:
https://github.com/kubernetes/kops/blob/0b0932b493bace5e695760bf5e5e26c82b560a76/pkg/pki/csr.go#L94

justinsb commented on upup/pkg/fi/fitasks/keypair.go:
```diff
@@ -180,15 +185,22 @@ func (_ *Keypair) Render(c *fi.Context, a, e, changes *Keypair) error {
+		} else if a.KeypairType == "" {
+			// The a.KeysetAPIVersion will == "" when the keyset is a legacy keyset and does not have a keypair.yaml
+			// file store in the state store
+			createCertificate = true
```
Why are you setting createCertificate here?

chrislovecnm replied:
I can test without it, but I think my testing did not create the keypair.yaml.

justinsb replied:
As we discussed, pretty sure this shouldn't drive creation of a new certificate, but should instead drive a new `UpdateStoredVersion` or similar call

justinsb commented on pkg/model/pki.go:
```diff
@@ -34,6 +34,9 @@ type PKIModelBuilder struct {
 
 var _ fi.ModelBuilder = &PKIModelBuilder{}
 
+// keypairType is used to differentiate the old secretes with the newer versions
```
typo: secretes

chrislovecnm replied:
we need the spelling tool

justinsb commented on pkg/model/pki.go:
```diff
@@ -34,6 +34,9 @@ type PKIModelBuilder struct {
 var _ fi.ModelBuilder = &PKIModelBuilder{}
 
+// keypairType is used to differentiate the old secretes with the newer versions
+const keypairType = "Keypair"
```
What is the other type?  I think maybe we have a field Format or "APIVersion", and we have "legacy" & "kops/v1alpha2" as that most closely matches what we're actually doing I think.

Although the obvious problem with this is that I don't know how easy it is to get the stored version from the apimachinery.  But I'm OK with cheating at the moment (assuming v1alpha2)

chrislovecnm replied:
This was the only thing in the API object that I could find that was actually set.  I tested the heck out of this, and the only thing I could figure out how to access is the object name Keypair.

justinsb commented on upup/pkg/fi/fitasks/keypair.go:
```diff
@@ -51,6 +51,9 @@ type Keypair struct {
 	Type string `json:"type"`
+	// KeypairType stores the api version of kops.Keyset.  We are storing this info in order to determine if kops
+	// is accessing legacy secrets that do not use keyset.yaml.
+	KeypairType string `json:"keypairType"`
```
I would call it APIVersion or Format.  And then I'd say "This lets us convert the stored keypair to our preferred version."  I don't think we should say "We are storing this info" because we're not storing it - tasks are ephemeral

chrislovecnm replied:
Sure ... renaming

justinsb commented on upup/pkg/fi/ca.go:
```diff
@@ -44,7 +44,7 @@ type KeystoreItem struct {
 	// FindKeypair finds a cert & private key, returning nil where either is not found
 	// (if the certificate is found but not keypair, that is not an error: only the cert will be returned)
-	FindKeypair(name string) (*pki.Certificate, *pki.PrivateKey, error)
+	FindKeypair(name string) (*pki.Certificate, *pki.PrivateKey, string, error)
```
I'm a little worried that we might need a version for both cert & privatekey, but we'll see...

chrislovecnm replied:
Well your test is not working ... see your PR for comments.

justinsb commented on upup/pkg/fi/k8sapi/k8s_keystore.go:
```diff
@@ -90,22 +90,23 @@ func (c *KubernetesKeystore) findSecret(id string) (*v1.Secret, error) {
 	}
 
-	return keypair.Certificate, keypair.PrivateKey, nil
+	// TODO: KubernetesKeystore does not seem to be using the newer version of keypairs.
```
Not sure what you mean.  keyset.yaml doesn't make sense here - keyset.yaml makes a filesystem behave more like an API server.   So we don't need keyset.yaml for an API-server-backed key store.

justinsb commented on upup/pkg/fi/vfs_castore.go:
```diff
@@ -233,27 +233,27 @@ func (c *VFSCAStore) parseKeysetYaml(data []byte) (*kops.Keyset, error) {
 // Returns (nil, nil) if the file is not found
 // Bundles avoid the need for a list-files permission, which can be tricky on e.g. GCE
-func (c *VFSCAStore) loadKeysetBundle(p vfs.Path) (*keyset, error) {
+func (c *VFSCAStore) loadKeysetBundle(p vfs.Path) (*keyset, string, error) {
```
You could put the stored version into the (internal) keyset type - should be simpler

chrislovecnm replied:
need more context

chrislovecnm commented on tests/keypair_test.go:
```diff
@@ -0,0 +1,267 @@
+// memfs://keystore/private/ca/keyset.yaml
+// memfs://keystore/private/kubelet/2.key
+// memfs://keystore/private/kubelet/4.key
+// memfs://keystore/private/kubelet/keyset.yaml
```
Test is failing and this is what I am getting :(

robinpercy commented on upup/pkg/fi/fitasks/keypair.go:
```diff
@@ -173,13 +181,18 @@ func (_ *Keypair) Render(c *fi.Context, a, e, changes *Keypair) error {
 			createCertificate = true
+			glog.V(8).Infof("creating certificate new Type")
 		} else {
 			glog.Warningf("Ignoring changes in key: %v", fi.DebugAsJsonString(changes))
```
@chrislovecnm As per our out-of-band convo, can we suppress this during an upgrade?

chrislovecnm replied:
Done

robinpercy commented:
I've run through a couple tests that create a cluster with kops 1.8, then use this patch to upgrade the cluster.  keyset.yaml is being created correctly.

chrislovecnm commented:
/assign @robinpercy 

PTAL

chrislovecnm commented:
/hold

chrislovecnm commented:
I have refactored some, and added a bunch of code level documentation.  Both @robinpercy and I have  tested and the upgrades are working well :)   When you upgrade to kops 1.9 from kops 1.8 and update and existing cluster built with kops 1.8 the new keypair.yaml files are created.

robinpercy commented:
/lgtm

chrislovecnm commented:
@justinsb I patched and modified your test PR here: https://github.com/kubernetes/kops/pull/4582/commits/ae759ce03fbceeb2b0e97487dddb8bc90019ffdc

Added set the `Format` member, and flipped the boolean value as we talked about.  I will drop the commit once you let me know that you have seen the changes.

The pass should be passing now.  PTAL

justinsb commented:
/lgtm

chrislovecnm commented:
/hold cancel
